### PR TITLE
UPDATE: [DEV-14211] update finetune to use tf-2.19 and native keras APIs

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

**Comment by madisonmay** [comment]
BugBot run

File: finetune/model.py [hunk 1]
```diff
@@ -1,386 +1,70 @@
...
+    
+    
+    class FinetuneModel(tf.keras.Model):
+        def __init__(self, *args, name="model", **kwargs):
+            super().__init__(*args, name=name, **kwargs)
+            self.featurizer = config.base_model.get_featurizer(encoder=encoder, config=config, name="featurizer")
+            self.target_block = ExtraScope(target_block, "target")
+            
+        def call(self, data, **kwargs):
+            print(data)
```
Comment (madisonmay) [inline finetune/model.py]: Remove me eventually

File: finetune/model.py [hunk 2]
```diff
@@ -1,386 +1,70 @@
...
-                max_length=params.max_length,
-                class_weights=weighted_tensor,
-                label_encoder=label_encoder,
+    
+    
+    class FinetuneModel(tf.keras.Model):
+        def __init__(self, *args, name="model", **kwargs):
+            super().__init__(*args, name=name, **kwargs)
+            self.featurizer = config.base_model.get_featurizer(encoder=encoder, config=config, name="featurizer")
+            self.target_block = ExtraScope(target_block, "target")
```
Comment (madisonmay) [inline finetune/model.py]: Honestly I like this solution a lot better than handling name mapping in saver.

File: finetune/base.py [hunk 3]
```diff
@@ -478,60 +318,44 @@ def _sort_by_length(
...
-                return list(outputs)
-            return outputs
+        output_list = []
+        for batch in input_fn():
+            preds = model(batch, training=False)
+            pred_numpy = {k: v.numpy() for k, v in preds.items()}
+            batch_size = batch["tokens"].shape[0]
+            # Items without a batch dim we include in all pred batches. In practice this is limited to just transition params
+            # but we can use it for anything we like going forwards.
+            not_batched = {k for k, v in pred_numpy.items() if v.shape[0] != batch_size}
```
Comment (madisonmay) [inline finetune/base.py]: Any chance of accidentally having shapes match batch size here?

File: finetune/saver.py [hunk 4]
```diff
@@ -1,177 +1,60 @@
...
     ):
         self.variable_transforms = variable_transforms or []
-        self.exclude_matches = exclude_matches
-        self.variables = None
         self.save_dtype = save_dtype
         if fallback_filename is not None:
             self.set_fallback(fallback_filename)
-        self.restart_global_step = restart_global_step
         self.permit_uninitialized = permit_uninitialized
-        self.add_tokens = add_tokens
```
Comment (madisonmay) [inline finetune/saver.py:174]: This was specific to LM / MLM, correct?

File: finetune/saver.py [hunk 5]
```diff
@@ -1,177 +1,60 @@
...
-            self._get_weights(session=run_context.session)
-
-    def end(self, session):
-        self.stop_if_no_metric_improvement_fn()
-        if not self.keep_best_model or self.saver.variables is None or self.get_current_weights:
-            self._get_weights(session=session)
+def set_weights(model: tf.keras.Model, weights: dict[str, np.ndarray], all_vars: dict[str, tf.Variable]):
+    for name, np_weight in weights.items():
+        if name.endswith("we:0") or "bert/embeddings/position_embedding" in name:
+            np_weight = np_weight[:all_vars[name].shape[0]]
```
Comment (madisonmay) [inline finetune/saver.py]: This is for reduced seq length position embeddings?

**Review by madisonmay — COMMENTED** [review]
Should we be using tf-keras or the standalone keras version (keras 3?)

- Seems like keep_best_model config should die
- Looks like gradient accumulation options should also probably die
- max_grad_norm has been removed -- I think we found we needed this in practice IIRC?
- Looks like we should remove any references to resume behavior / distribution strategy (or simply issue warnings where appropriate when those args are passed, since I guess we want to avoid breaking .load())
- New API for gradient checkpointing is super nice 

Would be nice to compile list of features that are deprecated after this change (and potentially add code in constructor that raises warnings when deprecated options are passed).

Think this is one that would also be worth stepping through on a call b/c the diff is pretty large.

File: finetune/base_models/bert/modeling.py [hunk 6]
```diff
@@ -15,114 +15,155 @@
...
-        return cls.from_dict(json.loads(text))
-
-    def to_dict(self):
-        """Serializes this instance to a Python dictionary."""
-        output = copy.deepcopy(self.__dict__)
-        return output
+        self.width = width
+        self.dense1 = tf.keras.layers.Dense(
+            width,
+            activation=tf.nn.relu,
```
Comment (benleetownsend) [inline finetune/base_models/bert/modeling.py]: Should be no activation here.

File: finetune/base.py [hunk 7]
```diff
@@ -478,60 +318,41 @@ def _sort_by_length(
...
+            # Items without a batch dim we include in all pred batches. In practice this is limited to just transition params
+            # but we include the shape check just to prevent errors if anything else sneaks through.
+            not_batched = {
+                k
+                for k, v in pred_numpy.items()
+                if v.shape[0] != batch_size or k == "transition_params"
+            }
+            for i in range(batch_size):
+                progress.update(i)
+                step_value = {
```
Comment (madisonmay) [inline finetune/base.py:411]: Rename so this doesn't sound like the step num

File: finetune/config.py [hunk 8]
```diff
@@ -206,44 +216,26 @@ def get_default_config():
...
         per_process_gpu_memory_fraction=None,
-        distribution_strategy="central_storage",
         xla=False,
         optimize_for="accuracy",
         sort_by_length=True,
         collapse_whitespace=False,
         permit_uninitialized=None,
-        max_training_hours=None,
-        feature_layer_num=None,
+        max_training_hours=None,  # TODO: maybe we want to keep this?
```
Comment (madisonmay) [inline finetune/config.py:225]: Remove me

File: finetune/input_pipeline.py [hunk 9]
```diff
@@ -230,33 +222,10 @@ def chunked_and_tokenized_dataset():
...
-            .shuffle(
-                self.config.shuffle_buffer_size,
-                seed=self.config.seed,
-                reshuffle_each_iteration=False,
-            )
-            .skip(self.config.val_size)
+        train_dataset = raw_dataset.shuffle(
+            self.config.shuffle_buffer_size,
+            seed=self.config.seed,
+            reshuffle_each_iteration=False,
```
Comment (madisonmay) [inline finetune/input_pipeline.py]: TODO  -- can we change this to True safely?

File: finetune/saver.py [hunk 10]
```diff
@@ -189,118 +91,62 @@ def fallback(self):
...
+    def initialize_model(self, model: tf.keras.Model):
+        transformed_weights: dict[str, np.ndarray] = {}
 
-            if self.variables is not None:
-                variables_sv = self.variables
+        if self.variables is not None:
+            variables_sv = self.variables
+        else:
+            variables_sv = dict()
+        print("Saved Model vars", variables_sv.keys())
```
Comment (madisonmay) [inline finetune/saver.py]: Some comments to remove

File: tests/test_backwards_compat.py [hunk 11]
```diff
@@ -53,3 +55,162 @@ def test_roberta_default_no_change(self):
...
+#     nested_assert_allclose(sequence_features, expected_sequence_features[1:2])
+
+
+@pytest.mark.parametrize("bundle_path", BUNDLES)
+def test_backwards_compat_extreme(bundle_path):
+    model = SequenceLabeler.load(bundle_path, key="model")
+    texts = SequenceLabeler.load(bundle_path, key="texts")
+    contexts = SequenceLabeler.load(bundle_path, key="contexts")
+
+    if False and model.config.float_16_predict:
```
Comment (madisonmay) [inline tests/test_backwards_compat.py]: Remove me

File: finetune/nn/target_blocks.py [hunk 12]
```diff
@@ -1,154 +1,32 @@
...
-        final_proj_w = tf.compat.v1.get_variable(
-            "dense/kernel",
-            [config.n_embed, config.n_embed],
-            initializer=tf.compat.v1.random_normal_initializer(
-                stddev=config.weight_stddev
-            ),
+    def build(self, input_shape):
+        self.w = self.add_weight(
+            shape=(input_shape[-1], self.n_targets),
+            initializer="random_normal",
```
Comment (madisonmay) [inline finetune/nn/target_blocks.py:16]: Double check config.weight_stddev

File: finetune/nn/target_blocks.py [hunk 13]
```diff
@@ -193,522 +71,194 @@ def _apply_multilabel_class_weight(
...
-):
-    def custom_grad_fn(op, g):
+# TODO; historically custom_gradient had very poor performance compared to using a defun - double check this isn't still the case.
+@tf.custom_gradient
+def class_reweighted_grad(logits, class_weights, norm_grads_multiplier):
+    def custom_grad_fn(g):
         new_g = g * class_weights
-        if norm_grads:
-            ratio = tf.math.divide_no_nan(tf.norm(g), tf.norm(new_g))
+        # This gets really badly autographed so we just need to use a float to cover these cases for now.
```
Comment (madisonmay) [inline finetune/nn/target_blocks.py:146]: Check perf w/o class weights

File: finetune/nn/target_blocks.py [hunk 14]
```diff
@@ -193,522 +71,194 @@ def _apply_multilabel_class_weight(
...
+        ) * norm_grads_multiplier + (1 - norm_grads_multiplier)
+        return [new_g * ratio, None, None]
+
+    return tf.identity(logits), custom_grad_fn
+
+class SequenceLabelerAttn(tf.keras.layers.Layer):
+    def __init__(self, n_targets, num_heads, name="seq_lab_attn", dtype=tf.float32, **kwargs):
+        super().__init__(name=name, **kwargs, dtype=dtype)
+        self.dense = tf.keras.layers.Dense(n_targets, name="dense", dtype=dtype)
+        self.block = Attn(num_heads=num_heads, attn_pdrop=0.2, mask=False, name="seq_label_attn", dtype=dtype)
```
Comment (madisonmay) [inline finetune/nn/target_blocks.py]: I think we have config vars for attn_pdrop

File: finetune/nn/target_blocks.py [hunk 15]
```diff
@@ -193,522 +71,194 @@ def _apply_multilabel_class_weight(
...
-                    tf.tile(a, [1, length, 1, 1]),
-                    tf.tile(b, [1, 1, length, 1]),
-                    # TODO: Think about using prediction as a feature for associations.
-                ],
-                axis=-1,
-            )
-            associations_flat = tf.compat.v1.layers.dense(
-                tf.reshape(features, shape=[-1, nx * 4]), num_associations
+    def compute_loss(self, layer_output, targets, class_weights):
+        # For some reason, all finetune targets are floats. I think we get more type flexibility
```
Comment (madisonmay) [inline finetune/nn/target_blocks.py]: Drop comment

File: finetune/base_models/bert/modeling.py [hunk 16]
```diff
@@ -148,185 +209,120 @@ class _BertModel(object):
...
+            batch_size=batch_size,
+            seq_length=seq_length,
+            attention_mask=attention_mask,
+        )
 
-    def get_embedding_table(self):
-        return self.embedding_table
+        output_shape = tf.shape(sequence_output)
 
+        def first_token():
```
Comment (madisonmay) [inline finetune/base_models/bert/modeling.py]: Is this still necessary?
